nrf52840-hal/src/gpio.rs

```diff
@@ -0,0 +1,339 @@
+                        // NOTE(unsafe) atomic write to a stateless register - TODO(AJM) verify?
+                        // TODO - I wish I could do something like `.pins$i()`...
+                        unsafe { (*$PX::ptr()).outclr.write(|w| w.bits(1u32 << $i)); }
+                    }
+                }
+
+                impl<MODE> StatefulOutputPin for $PXi<Output<MODE>> {
+                    /// Is the output pin set as high?
+                    fn is_set_high(&self) -> bool {
+                        !self.is_set_low()
+                    }
+
+                    /// Is the output pin set as low?
+                    fn is_set_low(&self) -> bool {
+                        // NOTE(unsafe) atomic read with no side effects - TODO(AJM) verify?
+                        // TODO - I wish I could do something like `.pins$i()`...
+                        unsafe { ((*$PX::ptr()).out.read().bits() & (1 << $i)) == 0 }
+                    }
+                }
+            )+
+        }
+    }
+}
+
+// ===========================================================================
+// Definition of all the items used by the macros above.
+// ===========================================================================
+gpio!(P0, p0, P0_Pin [
+    P0_0:  (p0_0,  0,  Input<Floating>),
+    P0_1:  (p0_1,  1,  Input<Floating>),
+    P0_2:  (p0_2,  2,  Input<Floating>),
+    P0_3:  (p0_3,  3,  Input<Floating>),
+    P0_4:  (p0_4,  4,  Input<Floating>),
+    P0_5:  (p0_5,  5,  Input<Floating>),
+    P0_6:  (p0_6,  6,  Input<Floating>),
+    P0_7:  (p0_7,  7,  Input<Floating>),
+    P0_8:  (p0_8,  8,  Input<Floating>),
+    P0_9:  (p0_9,  9,  Input<Floating>),
+    P0_10: (p0_10, 10, Input<Floating>),
+    P0_11: (p0_11, 11, Input<Floating>),
+    P0_12: (p0_12, 12, Input<Floating>),
+    P0_13: (p0_13, 13, Input<Floating>),
+    P0_14: (p0_14, 14, Input<Floating>),
+    P0_15: (p0_15, 15, Input<Floating>),
+    P0_16: (p0_16, 16, Input<Floating>),
+    P0_17: (p0_17, 17, Input<Floating>),
+    P0_18: (p0_18, 18, Input<Floating>),
+    P0_19: (p0_19, 19, Input<Floating>),
+    P0_20: (p0_20, 20, Input<Floating>),
+    P0_21: (p0_21, 21, Input<Floating>),
+    P0_22: (p0_22, 22, Input<Floating>),
+    P0_23: (p0_23, 23, Input<Floating>),
+    P0_24: (p0_24, 24, Input<Floating>),
+    P0_25: (p0_25, 25, Input<Floating>),
+    P0_26: (p0_26, 26, Input<Floating>),
+    P0_27: (p0_27, 27, Input<Floating>),
+    P0_28: (p0_28, 28, Input<Floating>),
+    P0_29: (p0_29, 29, Input<Floating>),
+    P0_30: (p0_30, 30, Input<Floating>),
+    P0_31: (p0_31, 31, Input<Floating>),
+
+]);
+
+// ===========================================================================
+// The macro doesn't seem to work for P1  as it has a different struct but uses
+// the same module. Not sure the best way to deal with P1 only implementing half the pins
+// either
+// ===========================================================================
+//gpio!(P1, p0, P1_Pin [
```

Comment:
Once https://github.com/jamesmunns/nrf52-hal/issues/8 is a bit further along, I think this will get a bit easier